cluster-app-values-schema/README.md

```diff
@@ -0,0 +1,373 @@
+
+Cluster apps are apps which are used to configure and provision clusters based on Cluster API in the Giant Swarm platform. Like all helm charts, these apps are configured using so-called values YAML files, which are passed to the app as a ConfigMap for cluster creation. The allowed structure of the values YAML is defined by the app's values schema.
+
+This RFC defines basic requirements for all cluster apps provided by Giant Swarm.
+
+## Overview
+
+- [R1: JSON Schema dialect (draft 2020-12) must be specified](#r1)
+- [R2: A single type must be declared](#r2)
+- [R3: Schema must explicitly cover all allowed properties](#r3)
+- [R4: Array item schema must be defined](#r4)
+- [R5: Properties must have a title](#r5)
+- [R6: Properties should have descriptions](#r6)
+- [R7: Properties should provide examples](#r7)
+- [R8: Constrain values as much as possible](#r8)
+- [R9: Required properties must be marked as such](#r9)
+- [R10: Use `anyOf` and `oneOf` only for specific purposes](#r10)
+- [R11: Use `deprecated` to phase out properties](#r11)
+- [R12: Provide valid string values where possible](#r12)
+- [R13: Avoid recursion](#r13)
+- [R14: Avoid logical constructs using `if`, `then`, `else`](#r14)
+- [R15: Avoid `unevaluatedProperties`, `unevaluatedItems`](#r15)
+- [R16: Array items and tuple validation](#r16)
+- [R17: Common schema structure](#r17)
+
+## Background
+
+TODO: Some more info regarding the rationale.
+
+- In-place documentation of properties
+- Better validation to catch configuration errors before they are applied
+- Generated entry forms
+- Generated information display
+- Enabling processes and communication about explicit contracts instead of implicit behaviour
+
+## Definitions
+
+- **Cluster app**: App (as defined by the Giant Swarm app platform) that allows for provisioning a cluster in a Giant Swarm installation.
+- **JSON Schema**: declarative language that allows to annotate and validate documents that are compatible with JSON.
+- The key words "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD", "SHOULD NOT", "RECOMMENDED",  "MAY", and "OPTIONAL" in this document are to be interpreted as described in [RFC 2119](https://www.rfc-editor.org/rfc/rfc2119.html)
+
+## Requirements
+
+Requirements carry identifiers with prefix `R` and a unique number, for easy referencing.
+
+### R1: JSON Schema dialect (draft 2020-12) must be specified {#r1}
+
+Each cluster app's values schema file MUST specify the schema dialect (also called draft) used via the `$schema` keyword on the top level, with the draft URI as a value.
+
+The draft URI MUST be `https://json-schema.org/draft/2020-12/schema`.
+
+Example:
+
+```json
+{
+  "$schema": "https://json-schema.org/draft/2020-12/schema",
+  ...
+}
+```
+
+### R2: A single type must be declared {#r2}
+
+For each property, including the root level schema, the `type` keyword MUST be present, and there MUST be a single value for the `type` keyword.
```

Comment:
I think we don't need `type`, when `$ref` is specified. For example:

```json
{
  "$schema": "https://json-schema.org/draft/2020-12/schema",
  "properties": {
    "address": {
      "$ref": "address.json"
    },
  "type": "object"
}

```
 Here I would expect that we don't require a type for `address`.

Reply:
Yes, true. Do you think this should be mentioned? To me that's covered by the definition of `$ref` in JSON Schema.

Reply:
This is important for the implementation of this rule in schemalint. JSON Schema does not forbid using `type` next to `$ref`. However, I think leaving out `type` when `$ref` is present is so intuitive, that we don't necessarily need to mention it here.